### PR TITLE
Fix #11802: Don't set edge traversability bit on aqueducts crossing that same edge side

### LINKED ISSUE #11802: [Bug]: Incorrect water region connectivity check with aqueduct on an edge
### Version of OpenTTD

20240116-master-g2efd897691

### Expected result

No connectivity!

Something like this should do the check the right way:
https://github.com/OpenTTD/OpenTTD/blob/2efd8976914d0a669656143a2cd20426df0b9901/src/script/api/script_marine.cpp#L59-L77

### Actual result

Captura de ecrã 2024-01-16 173937

### Steps to reproduce

https://github.com/OpenTTD/OpenTTD/blob/2efd8976914d0a669656143a2cd20426df0b9901/src/pathfinder/water_regions.cpp#L294-L297

### PR DESCRIPTION
**Pull request by SamuXarick**
## Motivation / Problem
#11802 - Neighbouring regions on the same x_or_y may indeed have the traversability bits sets, but that doesn't necessarily mean ships can traverse their edges.

## Description
Actually check if water regions are reachable.

Don't set edge traversability bit on aqueducts crossing that same edge side.

Closes #11802.

## Limitations

Almost same performance. What used to be 3000 us on a 4096x4096 map is now 3100 us, even without aqueducts.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/pathfinder/water_regions.cpp [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #include "tunnelbridge_map.h"
 #include "follow_track.hpp"
 #include "ship.h"
+#include "script/api/script_marine.hpp"
```
Comment (PeterN) [inline src/pathfinder/water_regions.cpp]: Pathfinder shouldn't be using script API methods...

**Comment by SamuXarick** [comment]
I believe this is now ready.

File: src/pathfinder/water_regions.cpp [hunk 2]
```diff
@@ -173,10 +173,25 @@ class WaterRegion
 		const int top_x = TileX(tile_area.tile);
 		const int top_y = TileY(tile_area.tile);
 		for (int i = 0; i < WATER_REGION_EDGE_LENGTH; ++i) {
-			if (GetWaterTracks(TileXY(top_x + i, top_y)) & TRACK_BIT_3WAY_NW) SetBit(this->edge_traversability_bits[DIAGDIR_NW], i); // NW edge
-			if (GetWaterTracks(TileXY(top_x + i, top_y + WATER_REGION_EDGE_LENGTH - 1)) & TRACK_BIT_3WAY_SE) SetBit(this->edge_traversability_bits[DIAGDIR_SE], i); // SE edge
-			if (GetWaterTracks(TileXY(top_x, top_y + i)) & TRACK_BIT_3WAY_NE) SetBit(this->edge_traversability_bits[DIAGDIR_NE], i); // NE edge
-			if (GetWaterTracks(TileXY(top_x + WATER_REGION_EDGE_LENGTH - 1, top_y + i)) & TRACK_BIT_3WAY_SW) SetBit(this->edge_traversability_bits[DIAGDIR_SW], i); // SW edge
+			TileIndex edge_tile = TileXY(top_x + i, top_y); // NW edge
+			if (GetWaterTracks(edge_tile) & TRACK_BIT_3WAY_NW && (!IsAqueductTile(edge_tile) || GetTunnelBridgeDirection(edge_tile) != DIAGDIR_NW)) {
```
Comment (glx22) [inline src/pathfinder/water_regions.cpp]: For readability:
```suggestion
			if ((GetWaterTracks(edge_tile) & TRACK_BIT_3WAY_NW) != 0 && (!IsAqueductTile(edge_tile) || GetTunnelBridgeDirection(edge_tile) != DIAGDIR_NW)) {
```
And the same for the other.

**Comment by SamuXarick** [comment]
Superseded by [#11986](https://github.com/OpenTTD/OpenTTD/pull/11986)

**Pull request closed without merging**